### PR TITLE
Refactor as an app-centric loopback component

### PR DESCRIPTION
**Pull request by mrfelton**
Hi @slively 
We have forked the loopback-db-migrate package and given it a fairly major overhaul and refactor. I wanted to reach out to you to get your thoughts on our version of the package and wether you feel it could be merged into your repo or wether you would prefer that we release as a separate project.

The main changes are:
- Refactor as a loopback component
- Pass `app` to migration handlers instead of a specific datasource
- Addition of unit tests
- Move migration functions to static methods on the Migration model
- Support promises
- Print migration timings through the process and as a summary at the end
- Use debug module to provide debugging information
- Ability to rollback to specific migration steps
- Ability to rollback a single failed migration
- Addition of Migration Map model to facilitate complex migration mappings
- Emit events on completion or error
- Addition of jshint and jscsrc
- Code style standardisation

Our fork currently does not have the cli component, although we would like to add one.

### CODE REVIEW

**Comment by doublemarked** [comment]
Cool!

**Comment by slively** [comment]
@mrfelton wow amazing job. How do you run migrations with this system? I use npm and cli interfaces for my builds and it's not entirely clear how I would run migrations for an environment or run rollbacks.

**Comment by mrfelton** [comment]
@slively  Currently, we both run programatically via a bootscript (based on starting the app with `MIGRATEDB=1`) and via a REST endpoint that we have added to a `System` endpoint that we have on our app that allows us to  `GET /api/v1/System/migrate?to=x` and `GET /api/v1/System/rollback?to=x`.

eg System model methods:

``` javascript
  /**
   * Remote Method: Run pending migrations.
   *
   * @param {Function} [cb] Callback function.
   */
  System.migrate = function (to, cb) {
    to = to || '';
    cb = cb || utils.createPromiseCallback();
    assert(typeof to === 'string', 'The to argument must be a string, not ' + typeof to);
    assert(typeof cb === 'function', 'The cb argument must be a function, not ' + typeof cb);
    System._runMigration('up', to, cb);
    return cb.promise;
  };

  /**
   * Remote Method: Rollback migrations.
   *
   * @param {Function} [cb] Callback function.
   */
  System.rollback = function (to, cb) {
    to = to || '';
    cb = cb || utils.createPromiseCallback();
    assert(typeof to === 'string', 'The to argument must be a string, not ' + typeof to);
    assert(typeof cb === 'function', 'The cb argument must be a function, not ' + typeof cb);
    System._runMigration('down', to, cb);
    return cb.promise;
  };

  /**
   * Helper Method: Run migrations or rollbacks.
   *
   * @param {Function} [cb] Callback function.
   */
  System._runMigration = function (direction, to, cb) {
    cb = cb || utils.createPromiseCallback();
    assert(typeof direction === 'string', 'The direction argument must be a string, not ' + typeof direction);
    assert(typeof to === 'string', 'The to argument must be a string, not ' + typeof to);
    assert(typeof cb === 'function', 'The cb argument must be a function, not ' + typeof cb);
    var promise = System.app.models.Setting.get('migrating')
      .then(function (migrating) {
        if (migrating && migrating.value) {
          return promise.cancel();
        }
        logger.info('Starting migration.');
        return System.app.models.Setting.set('migrating', true);
      })
      .then(function () {
        // Run the migration in the background.
        System.app.models.Migration.migrate(direction, to);
        cb();
      })
      .catch(Promise.CancellationError, function () {
        var err = new Error(util.format('Unable to migrate %s: already running.', direction));
        logger.info(err.message);
        return cb(err);
      })
      .catch(function (err) {
        logger.error(err);
      })
      .cancellable();
    return cb.promise;
  };
```

eg boot script:

``` javascript
module.exports = function (app) {

  // Initialise the migrate component.
  var migrate = require('loopback-db-migrate');
  migrate(app, {
    log: logger
  });

  // Only run if asked to
  if (!conf.get('MIGRATEDB')) {
    return;
  }

  // Trigger the migration scripts to run once the app has fully started.
  app.once('started', function () {
    return app.models.System.migrate('up');
  });

};
```

Although I'm thinking of moving those REST endpoints directly to the Migration model and including them in the loopback component so that you can run `GET /api/v1/Migration/migrate?to=x` and `GET /api/v1/Migration/rollback?to=x` out of the box. I also would like to get the CLI part back in there so that it's possible to run as a standalone script although I suspect the use case for that is less useful.

**Comment by doublemarked** [comment]
FYI - at least for us, cli interface is far preferable to REST endpoints. The cli is closer to how we do our other administration. And also, honestly, I am not comfortable exposing such routes from a 3rd party without having done much more careful analysis of how they work and evaluating the risks. 

Also, lastly, it feels odd to me that the app would launch in a state under which the databases are not migrated. This means that the LB engine is up with new code but the database structure is potentially from a previous release. In the time between when the app launches and the migration occurs clients hitting the system are potentially causing all sorts of mayhem. To reconcile this you're going to have to shutdown the HTTP gateway or something while you deal with the LB migration...sounds like bad voodoo. Better to migrate the database prior to allowing the app to accept remote connections at all.

**Comment by mrfelton** [comment]
> FYI - at least for us, cli interface is far preferable to REST endpoints. The cli is closer to how we do our other administration.

Understood. Wrapping the migrate and rollback methods up in a cli interface should be pretty simple.

> And also, honestly, I am not comfortable exposing such routes from a 3rd party without having done much more careful analysis of how they work and evaluating the risks.

Any REST endpoints would be blocked by ACLs by default leaving it up to the application owner to decide if it makes sense to open them up and to who.

> Also, lastly, it feels odd to me that the app would launch in a state under which the databases are not migrated. This means that the LB engine is up with new code but the database structure is potentially from a previous release. In the time between when the app launches and the migration occurs clients hitting the system are potentially causing all sorts of mayhem. To reconcile this you're going to have to shutdown the HTTP gateway or something while you deal with the LB migration...sounds like bad voodoo. Better to migrate the database prior to allowing the app to accept remote connections at all.

I think this is where things conceptually differ somewhat. I really don't see this as a solely database migration framework although it can of course be used for that. This is one of the reasons what I don't pass a datasource to the migration handlers but instead pass the app object. This is an app migration framework.

Applications often need to be migrated from one state to another as part of an ongoing release/upgrade process. Those updates can include but are not limited to database schema updates, data manipulation, calls to connected systems etc and many other things.

Its perfectly feasible to have a system online whilst updates are being performed, depending on the nature of those updates and it should be upto the application owner to decide how and when to hander their updates, be it by CLI, REST, automated boot process or some other means.

In our particular use case, we would typically flip the system into a 'maintenance mode' just prior to running updates to avid multi-application-state conflicts. The updates themselves would be be triggered by a management interface that uses the REST endpoints to run or rollback migrations. Code can be clever enough to know wether there are pending updates and to restrict usage of particular features if the app has not yet been migrated to the expect state for their usage.

In summary, my goal is to provide a robust framework that can run multiple types of application level updates by multiple means, without making assumptions about how or when those updates should be run.

**Comment by slively** [comment]
To put the debate of cli, boot script, and rest api aside, and in an effort to keep this project as small and simple as possible, I think running migrations in a boot script & through a REST api would be good as separate projects that could depend on this project to provide that functionality. Because you've created this as a loopback module it becomes nicely re-usable to create new use-cases on top of it.

To approve this I think it would need to include a CLI comparable to the original, drop any documentation relating to running it as part of a boot script, and keep the ability to have migrations for multiple datasources (currently your model only has name/runDtTm and no datasource).

**Comment by doublemarked** [comment]
@mrfelton I see, thank you for the more thorough description of your design goals. This is interesting, and I see that you're working towards something novel. 

> Applications often need to be migrated from one state to another as part of an ongoing release/upgrade process. Those updates can include but are not limited to database schema updates, data manipulation, calls to connected systems etc and many other things.

This is true, however - the first two items in your list (schema updates, data changes) are by and far the most commonly needed functionality for the majority of projects out there, and almost always these data migrations are to keep data state in sync with code state. Currently your PR seems to move in an interesting direction but may be overlooking what most people are seeking when finding `loopback-db-migrate`.

> In our particular use case, we would typically flip the system into a 'maintenance mode' just prior to running updates to avid multi-application-state conflicts. The updates themselves would be be triggered by a management interface that uses the REST endpoints to run or rollback migrations. 

That's interesting, and I agree that these scenarios come up sometimes. But it's still unclear to me how your approach deals with a more common schema change/data update, without ensuring the active code can operate with either schema/data state.

> Code can be clever enough to know wether there are pending updates and to restrict usage of particular features if the app has not yet been migrated to the expect state for their usage.

Ah, from this comment I guess you're thinking for code to able to operate with either schema state. Would this not be a non-trivial change to how we tend to design code? Do you have a concrete example of when/how you've taken this approach?

**Comment by listepo** [comment]
+1

**Comment by listepo** [comment]
@slively when you accept this PR?

**Comment by mrfelton** [comment]
Based on the conversation to date, I am leaning towards releasing my version of the package under the name loopback-component-migrate as it seems to me that my goals for the project outstretch the more limited use case that currently exists in loopback-db-migrate. I really don't see this as a database migration framework, but as a wider application migration framework. I also think that loopback-component-migrate is a more fitting name for the project.

**Comment by listepo** [comment]
@mrfelton, loopback-component-migrate :+1:

**Comment by listepo** [comment]
@mrfelton when you publish loopback-component-migrate?

**Comment by listepo** [comment]
@mrfelton I can do it

**Comment by mrfelton** [comment]
@listepo I have created the repo at https://github.com/fullcube/loopback-component-migrate. Just need to do some mass function renaming before I publish on npmjs. I'll happily accept patches :)

**Comment by listepo** [comment]
@mrfelton :+1:

**Comment by listepo** [comment]
@mrfelton what is the meaning of this PR if there https://github.com/fullcube/loopback-component-migrate?

**Comment by mrfelton** [comment]
@listepo I don't think there is any meaning any more as we decided to separate out into a separate project. Thanks for reminding me this PR was still open. Will close.
